bebraven/platform
# Add Honeycomb instrumentation for Javascript

Task: https://app.asana.com/0/1174274412967132/1188732532251833

**Note**: this PR should be reviewed alongside [this PR](https://github.com/bebraven/boomerang/pull/1). They are dependent on each other.

**Release Note**: just make sure and test that a basic trace comes through in prod since there is prod vs dev code.


Adds support for [Boomerang](https://github.com/akamai/boomerang) as a tool to auto-instrument our Javascript. Honeycomb doesn't directly support auto-instrumented browser JS, so we proxy the Boomerang "beacons" server side, turn them into "spans", do some translation of Boomerang fields, and send the events off the Honeycomb. See: https://docs.honeycomb.io/getting-data-in/javascript/browser-js/

#### Some notes:
1. We want the Honeycomb spans to show up in traces for the original controller that loaded the JS, not the proxy controller. We use a slightly undocumented `X-Honeycomb-Trace` header, set using middleware, to accomplish this.
2. The concept of a Honeycomb span doesn't directly translate to a Boomerang beacon. We add a wrapper class in JS to make it easier to ensure that Boomerang beacons translate into Honeycomb spans when we proxy them. 
3. The Boomerang fields are non-intuitive. We translate some standard ones into the Rails equivalent to make it easier to query for and analyze.
4. The loading of Boomerang should happen as early as possible but not interfere with the page load. This was hard to get right, but it turns out they have a [loader snippet](https://developer.akamai.com/tools/boomerang/docs/tutorial-loader-snippet.html) showing how to do it which we only needed to tweak slightly.
5. This depends on our custom built Boomerang JS. [See here for what we've customized](https://github.com/bebraven/boomerang/pull/1). We host the compiled version of this on AWS S3. This is how Boomerang recommends you do it. It can't be imported as an ES6 module b/c it doesn't comply with strict mode.

### Test Plan:
- Load a project and make sure the traces look good in Chrome, Chrome incognito, and Firefox (and IE 11 if you have time). On Honeycomb, just query for `trace.trace_id` output in the logs and then click the little icon shown in the second screenshot:
<img src="https://user-images.githubusercontent.com/5596986/92593240-c8f44f00-f26e-11ea-8691-49e61e59dc7e.png" width="800"> 
<img src="https://user-images.githubusercontent.com/5596986/92593361-fe993800-f26e-11ea-9028-c86469165285.png" width="120">

- Do something to cause the JS to throw an error and make sure the stacktrace comes through in the fields.

- Try it out both with the Boomerang JS cached by the browser and with it not cached. Make sure the traces look good. I use Chrome inspector to turn off browser caching to try this. The time it takes to download was giving me false positives on things working properly before I found their loader snippet.

- `yarn test`

- `bundle exec rspec`

### Example trace of loading a project with pre-populated answers:
<img src="https://user-images.githubusercontent.com/5596986/92491427-39906280-f1c0-11ea-95f4-3f00b4b7a90e.png" width="800">

### Example trace of sending an xAPI statement to the LRS from a project:
<img src="https://user-images.githubusercontent.com/5596986/92701254-c0a71d00-f31d-11ea-9386-9925dc5fee87.png" width="800">

### Example trace if an error happens in sending xAPI statements:
<img src="https://user-images.githubusercontent.com/5596986/92491951-d2bf7900-f1c0-11ea-9d6b-d45bc6c8d25a.png" width="800">

### Example trace with an error that happens during page load wrapped in a try/catch to add an error message
<img src="https://user-images.githubusercontent.com/5596986/92720417-92800800-f332-11ea-8cf2-122af1276d62.png" width="800">
